nslsii/sync_redis/sync_redis.py

```diff
@@ -0,0 +1,139 @@
+
+    finally:
+        return proposal_data
+
+
+def authenticate(username):
+
+    auth_server = Server("dc2.bnl.gov", use_ssl=True)
+
+    try:
+        connection = Connection(
+            auth_server, user=f"BNL\\{username}", password=getpass("Password : "), authentication=NTLM, auto_bind=True, raise_exceptions=True
+        )
+        print(f"\nAuthenticated as : {connection.extend.standard.who_am_i()}")
+
+    except LDAPInvalidCredentialsResult:
+        raise RuntimeError(f"Invalid credentials for user '{username}'.") from None
+
+
+def should_they_be_here(username, new_data_session, beamline):
+
+    user_access_json = nslsii_api_client.get(f"/v1/data-session/{username}").json()
+
+    if "nsls2" in user_access_json["facility_all_access"]:
+        return True
+
+    elif beamline.lower() in user_access_json["beamline_all_access"]:
+        return True
+
+    elif new_data_session in user_access_json["data_sessions"]:
+        return True
+
+    return False
+
+
+class AuthorizationError(Exception): ...
+
+
+def switch_experiment(proposal_number, beamline, verbose=False, prefix=""):
+
+    redis_client = redis.Redis(host=f"info.{beamline.lower()}.nsls2.bnl.gov")
+
+    md = RedisJSONDict(redis_client=redis_client, prefix=prefix)
+
+    new_data_session = f"pass-{proposal_number}"
+    username = input("Username : ")
+
+    if (new_data_session == md.get("data_session")) and (username == md.get("username")):
+
+        warnings.warn(f"Experiment {new_data_session} was already started by the same user.")
+
+    else:
+
+        proposal_data = validate_proposal(new_data_session, beamline)
+        users = proposal_data.pop("users")
+
+        authenticate(username)
+
+        if not should_they_be_here(username, new_data_session, beamline):
+            raise AuthorizationError(f"User '{username}' is not allowed to take data on proposal {new_data_session}")
+
+        md["data_session"] = new_data_session
+        md["username"] = username
+        md["start_datetime"] = datetime.now().isoformat()
+        md["cycle"] = get_current_cycle()
+
+        print(f"Started experiment {new_data_session}.")
+        if verbose:
+            print(json.dumps(proposal_data, indent=2))
+            print("Users:")
+            print(users)
+
+
+def main():
+    # Used by the `sync-redis` command
+
+    parser = argparse.ArgumentParser(description="Start or switch beamline experiment and record it in Redis")
+    parser.add_argument("-b", "--beamline", dest="beamline", type=str, help="Which beamline (e.g. CHX)", required=True)
+    parser.add_argument("-p", "--proposal", dest="proposal", type=int, help="Which proposal (e.g. 123456)", required=True)
+    parser.add_argument("-v", "--verbose", action=argparse.BooleanOptionalAction)
+
+    args = parser.parse_args()
+    switch_experiment(proposal_number=args.proposal, beamline=args.beamline, verbose=args.verbose)
+
+
+# e.g. start_experiment(proposal_number=314062, beamline="chx")
```

Comment:
This line can be eliminated.

Reply:
ah yeah, thanks, it was! committed, but didn't push...🤦‍♂️